contracts/epochs/FlowDKG.cdc

```diff
@@ -100,6 +90,230 @@ access(all) contract FlowDKG {
         }
     }
 
+    // Checks whether the ResultSubmission constructor arguments constitute a valid empty submission.
+    // A valid empty submission has all fields nil, and is used when the submittor locally failed the DKG.
+    // All non-empty submissions must have all non-nil fields.
+    access(all) view fun isValidNilSubmission(groupPubKey: String?, pubKeys: [String]?, idMapping: {String:Int}?): Bool {
+        // If any fields are nil, then this represents a empty submission and all fields must be nil
+        if groupPubKey == nil && pubKeys == nil && idMapping == nil {
+            return true
+        }
+        // Otherwise, all fields must be non-nil
+        return groupPubKey != nil && pubKeys != nil && idMapping != nil 
+    }
+
+    // Checks the group public key (part of a ResultSubmission) for validity.
+    // A valid public key in this context is either: (1) a hex-encoded 96 bytes, or (2) nil.
+    access(all) view fun isValidGroupKey(_ groupKey: String?): Bool {
+         if groupKey == nil {
+            // This is a nil/empty submission (see isValidNilSubmission)
+            return true
+        }
+        return groupKey!.length == FlowDKG.submissionKeyLength
+    }
+
+    // Checks a list of participant public keys (part of a ResultSubmission) for validity.
+    // A valid public key in this context is either: (1) a hex-encoded 96 bytes, or (2) nil.
+    access(all) view fun isValidPubKeys(_ pubKeys: [String]?): Bool {
+        if pubKeys == nil {
+            // This is a nil/empty submission (see isValidNilSubmission)
+            return true
+        }
+        for key in pubKeys! {
+            // keys must be exactly 96 bytes
+             if key.length != FlowDKG.submissionKeyLength {
+                 return false
+             }
+        }
+        return true
+    }
+
+    // Checks that an id mapping (part of ResultSubmission) contains one entry per public key.
+    access(all) view fun isValidIDMapping(pubKeys: [String]?, idMapping: {String: Int}?): Bool {
+        if pubKeys == nil {
+            // This is a nil/empty submission (see isValidNilSubmission)
+            return true
+        }
+        return pubKeys!.length == idMapping!.keys.length
+    }
+
+    // ResultSubmission represents a result submission from one DKG participant.
+    // Each submission includes a group public key and an ordered list of participant public keys.
+    // A submission may be empty, in which case all fields are nil - this is used to represent a local DKG failure.
+    // All non-empty submission will have all non-nil, valid fields.
+    // By convention, all keys are encoded as lowercase hex strings, though it is not enforced here.
+    //
+    // INVARIANTS:
+    //  (1) either all fields are nil (empty submission) or no fields are nil
```

Comment:
Previously, the data model for submissions was `[String?]`, so some keys could be nil and some could be non-nil. If any keys were nil, we considered the submission invalid ("empty" in the terminology used in this PR). 

This PR changes the data model so that either all fields are nil, or none are, and enforces this in the constructor.